### PR TITLE
2.0 Rewrite.

### PR DESCRIPTION
**Pull request by maxehmookau**

### CODE REVIEW

File: app.rb [hunk 1]
```diff
@@ -3,46 +3,81 @@
...
-  {
-    token:      params[:token],
-    status:     pipe[0],
-    error:      pipe[1],
-    domain:     pipe[2],
-    subdomains: pipe[3].to_s.split(','),
-    message:    pipe[4]
-  }.to_json
+def params_valid?
+  !@request_payload["domains"].empty? && !@request_payload["heroku_app_name"].empty? && !@request_payload["zone"].empty?
```
Comment (samknight) [inline app.rb]: `.any?` might be more readable

```(@request_payload["domains"] + @request_payload["heroku_app_name"] + @request_payload["zone"]).any?``` might also work but might not be as obvious what it's doing

File: app.rb [hunk 2]
```diff
@@ -3,46 +3,81 @@
...
+#   }.to_json
+# end
+
 private
 
 def authenticate!
-  unless (params[:auth_token] == ENV['AUTH_TOKEN'])
-    halt 403, 'Not authenticated'
+  unless (@request_payload["auth_token"] == ENV['AUTH_TOKEN'])
+    halt 403
```
Comment (samknight) [inline app.rb]: ```halt 403 if @request_payload["auth_token"] != ENV['AUTH_TOKEN']```

File: lib/acme_client_registration.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  class NoEmailError < StandardError; end;
+
+  DEBUG_ENDPOINT = "https://acme-staging.api.letsencrypt.org/"
+  LIVE_ENDPOINT  = "https://acme-v01.api.letsencrypt.org/"
+
+  attr_reader :endpoint, :client
+
+  def initialize(options = {})
+    raise NoEmailError if ENV['CONTACT_EMAIL'].empty?
+    if options[:debug]
```
Comment (samknight) [inline lib/acme_client_registration.rb]: what's your feeling on the term debug? As mentioned in a different thread I feel like debug should be more verbose output with the live api. Test/Demo etc might be better terms/

File: app.rb (hunk 1 above)
Comment (maxehmookau) [inline app.rb]: I did that originally, but `.any?` appears to be an ActiveSupport thing, not an `Array` thing. :(

Comment (samknight) [inline app.rb]: It's Ruby 2.2 but understand if you want to support older versions

Comment (maxehmookau) [inline app.rb]: Hmm... nope. I'm aiming for 2.3 or above so it should work. I'll look more in to this!

**Comment by richardsimko** [comment]
Does this mean that 2.0 is released? :D

**Comment by maxehmookau** [comment]
Yup! Would love to get feedback on it. We're already working on a few new bits, first of all: Allowing cloudflare/heroku details to be passed through the API as params rather than environment variables.

https://github.com/substrakt/letsencrypt-heroku/releases/tag/2.0.0

**Comment by richardsimko** [comment]
Great! Will look at setting it up again during the week.

**Pull request merged**